netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -797,101 +901,227 @@ _net_ebpf_extension_connection_context_initialize(
 
-    NET_EBPF_EXT_RETURN_POINTER(net_ebpf_extension_connection_context_t*, connection_context);
+    return entry_found;
 }
 
-_Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_lock) static void _net_ebpf_ext_purge_lru_contexts_under_lock(
-    BOOLEAN delete_all)
+_Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_blocked_contexts
+                                   .lock) static void _net_ebpf_ext_purge_blocked_connect_contexts(bool delete_all)
 {
     uint64_t expiry_time = CONVERT_100NS_UNITS_TO_MS(KeQueryInterruptTime()) - EXPIRY_TIME;
 
-    LIST_ENTRY* list_entry = _net_ebpf_ext_connect_context_list.Blink;
-    while (list_entry != &_net_ebpf_ext_connect_context_list) {
+    // Free entries from the LRU list. These entries should also be removed from the table.
+    LIST_ENTRY* list_entry = _net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_lru_list.Blink;
+    while (list_entry != &_net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_lru_list) {
         net_ebpf_extension_connection_context_t* entry =
             CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
+        // Move pointer to next entry prior to removing the entry.
+        list_entry = list_entry->Blink;
+
         if (!delete_all && entry->timestamp > expiry_time) {
             break;
         }
-#pragma warning(suppress : 6001) /* list entry is non-null */
-        list_entry = list_entry->Blink;
-        RemoveEntryList(&entry->list_entry);
-
-        _net_ebpf_ext_connect_context_count--;
 
+#pragma warning(suppress : 6001) /* entry and list entry are non-null */
+        RemoveEntryList(&entry->list_entry);
+        BOOLEAN result =
+            RtlDeleteElementGenericTableAvl(&_net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_table, entry);
+        ebpf_assert(result);
+        _net_ebpf_ext_sock_addr_blocked_contexts.blocked_context_count--;
         NET_EBPF_EXT_LOG_MESSAGE_UINT64(
             NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
             NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
-            "_net_ebpf_ext_purge_lru_contexts_under_lock: Delete",
+            "_net_ebpf_ext_purge_block_connect_contexts: Delete",
             entry->transport_endpoint_handle);
+    }
+
+    // Free entries from low-memory list.
+    list_entry = _net_ebpf_ext_sock_addr_blocked_contexts.low_memory_blocked_context_list.Blink;
+    while (list_entry != &_net_ebpf_ext_sock_addr_blocked_contexts.low_memory_blocked_context_list) {
+        net_ebpf_extension_connection_context_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_connection_context_t, list_entry);
+        if (!delete_all && entry->timestamp > expiry_time) {
+            break;
+        }
+
+#pragma warning(suppress : 6001) /* entry and list entry are non-null */
+        list_entry = list_entry->Blink;
+        RemoveEntryList(&entry->list_entry);
 
-        ExFreePool(entry);
+        if (delete_all) {
```

Comment:
nit: if we have `delete_all` check here, I think the stale entries that are deleted from AVL tree will be leaked?

Reply:
This part of the code is only using the low_memory_free_context_list. If delete_all is true, then we are doing cleanup, so we'll want to free the entry. Else, we'll want to return it back to the low_memory_free_context_list.

For the above portion of the code, where we are using entries from the AVL tree, we'll be freeing whenever they are removed.